lib/tests/test_merge_trees.rs

```diff
@@ -645,3 +645,65 @@ fn test_simplify_conflict_after_resolving_parent() {
 
 // TODO: Add tests for simplification of multi-way conflicts. Both the content
 // and the executable bit need testing.
+
+#[test]
+fn test_rebase_on_lossy_merge() {
+    let settings = testutils::user_settings();
+    let test_repo = TestRepo::init();
+    let repo = &test_repo.repo;
+
+    // Test this rebase:
+    // D    foo=2          D'   foo=3
+    // |\                  |\
+    // | C  foo=2          | C' foo=3
+    // | |           =>    | |
+    // B |  foo=2          B |  foo=2
+    // |/                  |/
+    // A    foo=1          A    foo=1
+    // Commit D effectively discarded a change from "1" to "2", so the result should
+    // be "3".
```

Comment:
Somewhat optional: I would explain this in more detail. I don't think "the result should be 3", I think the result of a conflict or 3 is equally right and equally problematic (if we take `A + A - B = A` seriously), and it might change later. In other words, it depends on whether we think of `D` as the "empty commit" or as a snapshot that tries to follow normal patch theory rules (without `A + A - B = A`).

This is not necessarily blocking, I can write something up later if I remember to do it.

Reply:
In any case, good job finding this example, it's very illustrative as an example of A + A - B = A confusion.

Reply:
I've added more explanation to the code comment.

Reply:
The additional explanation doesn't help. Is the original `D` supposed to be "foo=3", so that your comment
> Commit D effectively discarded a change from "1" to "2"

actually makes sense, or should it be referring to `D'`?
Maybe instead of just saying "Test this rebase", you state what "this" is.

Reply:
`A` starts with contents "1" in `foo`. `B` and `C` both represent a patch that changes "1" to "2" (in `foo`). So the merge of them should strictly speaking be a conflict (and some VCSs like Darcs and Pijul would consider it a conflict). Still, the state in `D` is just "2" (not a conflict). That's what I mean when I said it effectively discards a change from "1" to "2".

I think it's clear enough for those likely to work on this code. We've talked about the conflict algebra often enough that I think they will understand.